File: crates/lumit-ui/src/icons.rs
//! Lumit's icons: the Iconoir set, rendered from an embedded icon font
//! (K-085, via the `iconflow` crate; Iconoir is MIT-licensed).
//!
//! In plain terms: instead of drawing every icon by hand from lines and curves
//! (the old way), Lumit now ships a professionally drawn icon family as a
//! small font file baked into the program. Each icon is a character in that
//! font, so it stays razor-sharp at any size and always takes the exact theme
//! colour we ask for — dimming on hover, turning accent when active — just
//! like text does. Emoji are still banned: every glyph here is a real icon
//! from one consistent set, never a character we hope the user's fonts carry.
//!
//! [`install`] must run once at startup (Theme::install_fonts does this)
//! before anything paints, or the icon font family won't exist.

use egui::{Align2, Color32, FontFamily, FontId, Painter, Pos2, Rect, RichText, Vec2};
use iconflow::{Pack, Size, Style};

/// One icon. Add a variant here and its Iconoir name in [`Icon::name`]; the
/// `every_icon_resolves` test fails on any name the pack doesn't carry.
#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum Icon {
    /// Selection arrow (the Select tool).
    Pointer,
    /// Pan the view (the Hand tool).
    Move,
    /// Mask/shape tool — rectangle.
    Rectangle,
    /// Mask/shape tool — ellipse.
    Ellipse,
    /// Mask/shape tool — star.
    Star,
    /// Pen (nib) tool.
    Pen,
    /// Transport: play.
    Play,
    /// Transport: pause.
    Pause,
    /// Closed padlock (aspect ratio locked).
    Lock,
    /// Open padlock (aspect ratio free).
    Unlock,
    /// Chain link (linked values, e.g. linked scale).
    Link,
    /// Folder (a project folder).
    Folder,
    /// Film (the "new composition" button).
    Film,
    /// Graph editor view: an animation curve with control points.
    GraphCurve,
    /// Layer/timeline view: stacked bars.
    TimelineBars,
    /// Node graph view (the future node system).
    Nodes,
    /// Footage item: a media clip.
    Footage,
    /// Composition item.
    Comp,
    /// Solid item: a filled block of colour.
    Solid,
    /// Sequence layer: clips cut back-to-back on a row.
    Sequence,
    /// Text layer.
    Text,
    /// Camera layer.
    Camera,
    /// Layer visibility switch: an eye.
    Eye,
    /// Hidden layer: an eye, closed (the visibility switch when off).
    EyeClosed,
    /// Audible layer: a speaker.
    Audio,
    /// Muted layer: a speaker, off.
    Mute,
    /// Pop a panel out into its own window.
    PopOut,
    /// Jump to the previous keyframe.
    PrevKeyframe,
    /// Jump to the next keyframe.
    NextKeyframe,
    /// Add a keyframe at the playhead.
    KeyframeAdd,
    /// A keyframe (outline diamond) — the navigator's add button when the
    /// playhead is between keys.
    Keyframe,
    /// A keyframe, filled — shown on the navigator's add/remove button when the
    /// playhead sits on a key (note 2.4).
    KeyframeFilled,
    /// The animate toggle on a property row.
    Stopwatch,
    /// Disclosure twirl, closed (points right).
    TwirlClosed,
    /// Disclosure twirl, open (points down).
    TwirlOpen,
    /// Collapse transformations (Precomp layers): the AE-style sunburst.
    Collapse,
    /// The Flow option on footage layers (K-088): optical-flow interpolation.
    Flow,
    /// The 3D column: a cube (the header glyph over the per-layer 3D switch).
    Cube3d,
    /// Snapping toggle in the timeline's bottom bar: a magnet.
    Magnet,
    /// Sample a colour from the Viewer: an eyedropper/pipette.
    Eyedropper,
    /// Reset a control to its default (EC4): a circular refresh arrow.
    Reset,
    /// Motion blur (the comp master and the per-layer switch): an arrow with
    /// speed lines — the closest Iconoir glyph to a motion smear (owner).
    MotionBlur,
}

impl Icon {
    /// Every variant, for exhaustive iteration (tests, palettes).
    pub const ALL: [Icon; 42] = [
        Icon::Pointer,
        Icon::Move,
        Icon::Rectangle,
        Icon::Ellipse,
        Icon::Star,
        Icon::Pen,
        Icon::Play,
        Icon::Pause,
        Icon::Lock,
        Icon::Unlock,
        Icon::Link,
        Icon::Folder,
        Icon::Film,
        Icon::GraphCurve,
        Icon::TimelineBars,
        Icon::Nodes,
        Icon::Footage,
        Icon::Comp,
        Icon::Solid,
        Icon::Sequence,
        Icon::Text,
        Icon::Camera,
        Icon::Eye,
        Icon::EyeClosed,
        Icon::Audio,
        Icon::Mute,
        Icon::PopOut,
        Icon::PrevKeyframe,
        Icon::NextKeyframe,
        Icon::KeyframeAdd,
        Icon::Keyframe,
        Icon::KeyframeFilled,
        Icon::Stopwatch,
        Icon::TwirlClosed,
        Icon::TwirlOpen,
        Icon::Collapse,
        Icon::Flow,
        Icon::Cube3d,
        Icon::Magnet,
        Icon::Eyedropper,
        Icon::Reset,
        Icon::MotionBlur,
    ];

    /// The Iconoir icon this variant renders.
    fn name(self) -> &'static str {
        match self {
            Icon::Pointer => "cursor-pointer",
            Icon::Move => "drag-hand-gesture",
            Icon::Rectangle => "square",
            Icon::Ellipse => "circle",
            Icon::Star => "star",
            Icon::Pen => "design-nib",
            Icon::Play => "play",
            Icon::Pause => "pause",
            Icon::Lock => "lock",
            Icon::Unlock => "lock-slash",
            Icon::Link => "link",
            Icon::Folder => "folder",
            Icon::Film => "movie",
            Icon::GraphCurve => "ease-curve-control-points",
            Icon::TimelineBars => "align-left",
            Icon::Nodes => "network",
            Icon::Footage => "media-video",
            Icon::Comp => "frame",
            Icon::Solid => "fill-color",
            Icon::Sequence => "view-columns-3",
            Icon::Text => "text",
            Icon::Camera => "video-camera",
            Icon::Eye => "eye",
            Icon::EyeClosed => "eye-closed",
            Icon::Audio => "sound-high",
            Icon::Mute => "sound-off",
            Icon::PopOut => "open-new-window",
            Icon::PrevKeyframe => "nav-arrow-left",
            Icon::NextKeyframe => "nav-arrow-right",
            Icon::KeyframeAdd => "keyframe-plus",
            Icon::Keyframe => "keyframe",
            Icon::KeyframeFilled => "keyframe",
            Icon::Stopwatch => "timer",
            Icon::TwirlClosed => "nav-arrow-right",
            Icon::TwirlOpen => "nav-arrow-down",
            Icon::Collapse => "flare",
            Icon::Flow => "wind",
            Icon::Cube3d => "cube",
            Icon::Magnet => "magnet",
            Icon::Eyedropper => "color-picker",
            Icon::Reset => "refresh-double",
            Icon::MotionBlur => "fast-arrow-right",
        }
    }

    /// The drawing style for this icon. Almost every glyph is the Regular
    /// (outline) weight; a few name a Filled variant of the same Iconoir icon so
    /// a toggle can swap outline↔filled to show state (e.g. a keyframe that is
    /// hollow between keys and solid on a key — note 2.4).
    fn style(self) -> Style {
        match self {
            Icon::KeyframeFilled => Style::Filled,
            _ => Style::Regular,
        }
    }

    /// The glyph and font family for this icon, or None if the pack lacks it
    /// (guarded by `every_icon_resolves`, so None never happens in practice —
    /// and the UI degrades to painting nothing rather than faulting).
    fn glyph(self) -> Option<(char, &'static str)> {
        let r = iconflow::try_icon(Pack::Iconoir, self.name(), self.style(), Size::Regular).ok()?;
        Some((char::from_u32(r.codepoint)?, r.family))
    }
}

/// Register the icon font into a set of font definitions. Must run before
/// anything paints an icon; `Theme::install_fonts` calls it at startup.
pub fn install(defs: &mut egui::FontDefinitions) {
    for font in iconflow::fonts() {
        defs.font_data.insert(
            font.family.to_owned(),
            std::sync::Arc::new(egui::FontData::from_static(font.bytes)),
        );
        defs.families
            .entry(FontFamily::Name(font.family.into()))
            .or_default()
            .insert(0, font.family.to_owned());
    }
}

/// Paint `icon` centred in `rect` in `color`. The glyph fills the smaller side
/// of the rect (Iconoir's own padding keeps strokes off the edge). `_width` is
/// kept for call-site compatibility with the old stroke-drawn icons.
pub fn paint(painter: &Painter, rect: Rect, icon: Icon, color: Color32, _width: f32) {
    let Some((glyph, family)) = icon.glyph() else {
        return;
    };
    let size = rect.width().min(rect.height());
    painter.text(
        rect.center(),
        Align2::CENTER_CENTER,
        glyph,
        FontId::new(size, FontFamily::Name(family.into())),
        color,
    );
}

/// The icon as inline rich text at `size` px, for `egui::Button`/`ui.label`.
/// No colour is set, so the widget's own state colouring applies (disabled
/// buttons dim their icon like they dim their text).
pub fn text(icon: Icon, size: f32) -> RichText {
    let Some((glyph, family)) = icon.glyph() else {
        return RichText::new("");
    };
    RichText::new(glyph.to_string()).font(FontId::new(size, FontFamily::Name(family.into())))
}

/// A disclosure twirl: points right when closed, down when open.
pub fn disclosure(painter: &Painter, rect: Rect, open: bool, color: Color32) {
    let icon = if open {
        Icon::TwirlOpen
    } else {
        Icon::TwirlClosed
    };
    paint(painter, rect, icon, color, 1.0);
}

/// The animate toggle on a property row: Iconoir's timer, sized from the old
/// stopwatch's radius. `animated` state is carried by the colour the caller
/// picks (accent when animated), so the glyph itself doesn't change.
pub fn stopwatch(painter: &Painter, center: Pos2, radius: f32, _animated: bool, color: Color32) {
    let side = radius * 2.0 + 4.0;
    paint(
        painter,
        Rect::from_center_size(center, Vec2::splat(side)),
        Icon::Stopwatch,
        color,
        1.0,
    );
}

/// A small downward caret marking a control as a dropdown.
pub fn caret_down(painter: &Painter, center: Pos2, color: Color32) {
    paint(
        painter,
        Rect::from_center_size(center, Vec2::splat(9.0)),
        Icon::TwirlOpen,
        color,
        1.0,
    );
}

#[cfg(test)]
#[allow(clippy::unwrap_used, clippy::expect_used, clippy::panic)]
mod tests {
    use super::*;

    /// Every variant's Iconoir name resolves in the embedded pack — a typo'd
    /// or removed icon name fails here, not silently in the UI.
    #[test]
    fn every_icon_resolves() {
        for icon in Icon::ALL {
            let r = iconflow::try_icon(Pack::Iconoir, icon.name(), Style::Regular, Size::Regular);
            assert!(r.is_ok(), "{icon:?} → {:?} does not resolve", icon.name());
            let r = r.unwrap();
            assert!(
                char::from_u32(r.codepoint).is_some(),
                "{icon:?} has an invalid codepoint"
            );
        }
    }

    /// Every icon paints without panicking once the fonts are installed
    /// (unknown font families make egui's layouter fault, so this also guards
    /// the install path).
    #[test]
    fn every_icon_paints() {
        let ctx = egui::Context::default();
        let mut defs = egui::FontDefinitions::default();
        install(&mut defs);
        ctx.set_fonts(defs);
        let _ = ctx.run(egui::RawInput::default(), |ctx| {
            egui::CentralPanel::default().show(ctx, |ui| {
                let painter = ui.painter().clone();
                for icon in Icon::ALL {
                    paint(
                        &painter,
                        Rect::from_min_size(Pos2::ZERO, Vec2::splat(16.0)),
                        icon,
                        Color32::WHITE,
                        1.5,
                    );
                }
            });
        });
    }
}

File: flutter_ui/lib/icons/icons.dart
// Lumit's icons: the Iconoir set (MIT), ported one-for-one from
// crates/lumit-ui/src/icons.rs (K-085). Same rules as the Rust side: every
// glyph is a real icon from one consistent set, emoji are banned, icons take
// the text colour of their state, and the motion-blur mark is drawn from the
// owner's artwork rather than looked up (Iconoir has no motion-blur glyph).

import 'package:flutter/widgets.dart';
import 'package:iconoir_flutter/regular/align_left.dart' as ic;
import 'package:iconoir_flutter/regular/circle.dart' as ic;
import 'package:iconoir_flutter/regular/color_picker.dart' as ic;
import 'package:iconoir_flutter/regular/cube.dart' as ic;
import 'package:iconoir_flutter/regular/cursor_pointer.dart' as ic;
import 'package:iconoir_flutter/regular/design_nib.dart' as ic;
import 'package:iconoir_flutter/regular/drag_hand_gesture.dart' as ic;
import 'package:iconoir_flutter/regular/ease_curve_control_points.dart' as ic;
import 'package:iconoir_flutter/regular/eye.dart' as ic;
import 'package:iconoir_flutter/regular/eye_closed.dart' as ic;
import 'package:iconoir_flutter/regular/fill_color.dart' as ic;
import 'package:iconoir_flutter/regular/flare.dart' as ic;
import 'package:iconoir_flutter/regular/folder.dart' as ic;
import 'package:iconoir_flutter/regular/frame.dart' as ic;
import 'package:iconoir_flutter/regular/fx.dart' as ic;
import 'package:iconoir_flutter/regular/keyframe.dart' as ic;
import 'package:iconoir_flutter/regular/keyframe_plus.dart' as ic;
import 'package:iconoir_flutter/regular/link.dart' as ic;
import 'package:iconoir_flutter/regular/link_xmark.dart' as ic;
import 'package:iconoir_flutter/regular/lock.dart' as ic;
import 'package:iconoir_flutter/regular/lock_slash.dart' as ic;
import 'package:iconoir_flutter/regular/magnet.dart' as ic;
import 'package:iconoir_flutter/regular/media_video.dart' as ic;
import 'package:iconoir_flutter/regular/movie.dart' as ic;
import 'package:iconoir_flutter/regular/nav_arrow_down.dart' as ic;
import 'package:iconoir_flutter/regular/nav_arrow_left.dart' as ic;
import 'package:iconoir_flutter/regular/nav_arrow_right.dart' as ic;
import 'package:iconoir_flutter/regular/network.dart' as ic;
import 'package:iconoir_flutter/regular/open_new_window.dart' as ic;
import 'package:iconoir_flutter/regular/pause.dart' as ic;
import 'package:iconoir_flutter/regular/play.dart' as ic;
import 'package:iconoir_flutter/regular/refresh_double.dart' as ic;
import 'package:iconoir_flutter/regular/sound_high.dart' as ic;
import 'package:iconoir_flutter/regular/sound_off.dart' as ic;
import 'package:iconoir_flutter/regular/square.dart' as ic;
import 'package:iconoir_flutter/regular/star.dart' as ic;
import 'package:iconoir_flutter/regular/text.dart' as ic;
import 'package:iconoir_flutter/regular/timer.dart' as ic;
import 'package:iconoir_flutter/regular/video_camera.dart' as ic;
import 'package:iconoir_flutter/regular/view_columns_3.dart' as ic;
import 'package:iconoir_flutter/regular/wind.dart' as ic;
import 'package:iconoir_flutter/solid/keyframe.dart' as ics;

/// One icon — the same 44 variants as the Rust `Icon` enum, same names.
enum LumitIcon {
  pointer,
  move,
  rectangle,
  ellipse,
  star,
  pen,
  play,
  pause,
  lock,
  unlock,
  link,
  unlink,
  folder,
  film,
  graphCurve,
  timelineBars,
  nodes,
  footage,
  comp,
  solid,
  sequence,
  text,
  camera,
  eye,
  eyeClosed,
  audio,
  mute,
  popOut,
  prevKeyframe,
  nextKeyframe,
  keyframeAdd,
  keyframe,
  keyframeFilled,
  stopwatch,
  twirlClosed,
  twirlOpen,
  collapse,
  flow,
  cube3d,
  magnet,
  eyedropper,
  reset,
  motionBlur,
  fx,
}

/// Build `icon` at `size` in `color`. The motion-blur mark is drawn, not
/// looked up, exactly as in the Rust frontend.
Widget lumitIcon(LumitIcon icon, {required double size, required Color color}) {
  if (icon == LumitIcon.motionBlur) {
    return CustomPaint(
      size: Size.square(size),
      painter: MotionBlurPainter(color),
    );
  }
  final w = _glyph(icon, color);
  return SizedBox(width: size, height: size, child: w);
}

Widget _glyph(LumitIcon icon, Color color) => switch (icon) {
      LumitIcon.pointer => ic.CursorPointer(color: color),
      LumitIcon.move => ic.DragHandGesture(color: color),
      LumitIcon.rectangle => ic.Square(color: color),
      LumitIcon.ellipse => ic.Circle(color: color),
      LumitIcon.star => ic.Star(color: color),
      LumitIcon.pen => ic.DesignNib(color: color),
      LumitIcon.play => ic.Play(color: color),
      LumitIcon.pause => ic.Pause(color: color),
      LumitIcon.lock => ic.Lock(color: color),
      LumitIcon.unlock => ic.LockSlash(color: color),
      LumitIcon.link => ic.Link(color: color),
      LumitIcon.unlink => ic.LinkXmark(color: color),
      LumitIcon.folder => ic.Folder(color: color),
      LumitIcon.film => ic.Movie(color: color),
      LumitIcon.graphCurve => ic.EaseCurveControlPoints(color: color),
      LumitIcon.timelineBars => ic.AlignLeft(color: color),
      LumitIcon.nodes => ic.Network(color: color),
      LumitIcon.footage => ic.MediaVideo(color: color),
      LumitIcon.comp => ic.Frame(color: color),
      LumitIcon.solid => ic.FillColor(color: color),
      LumitIcon.sequence => ic.ViewColumns3(color: color),
      LumitIcon.text => ic.Text(color: color),
      LumitIcon.camera => ic.VideoCamera(color: color),
      LumitIcon.eye => ic.Eye(color: color),
      LumitIcon.eyeClosed => ic.EyeClosed(color: color),
      LumitIcon.audio => ic.SoundHigh(color: color),
      LumitIcon.mute => ic.SoundOff(color: color),
      LumitIcon.popOut => ic.OpenNewWindow(color: color),
      LumitIcon.prevKeyframe => ic.NavArrowLeft(color: color),
      LumitIcon.nextKeyframe => ic.NavArrowRight(color: color),
      LumitIcon.keyframeAdd => ic.KeyframePlus(color: color),
      LumitIcon.keyframe => ic.Keyframe(color: color),
      LumitIcon.keyframeFilled => ics.KeyframeSolid(color: color),
      LumitIcon.stopwatch => ic.Timer(color: color),
      LumitIcon.twirlClosed => ic.NavArrowRight(color: color),
      LumitIcon.twirlOpen => ic.NavArrowDown(color: color),
      LumitIcon.collapse => ic.Flare(color: color),
      LumitIcon.flow => ic.Wind(color: color),
      LumitIcon.cube3d => ic.Cube(color: color),
      LumitIcon.magnet => ic.Magnet(color: color),
      LumitIcon.eyedropper => ic.ColorPicker(color: color),
      LumitIcon.reset => ic.RefreshDouble(color: color),
      LumitIcon.motionBlur => const SizedBox.shrink(), // handled above
      LumitIcon.fx => ic.Fx(color: color),
    };

/// The motion-blur mark: a ring with speed streaks running into it, from the
/// owner's artwork on a 24×24 grid — coordinates identical to the Rust
/// `draw_motion_blur` so the two frontends paint the same mark.
class MotionBlurPainter extends CustomPainter {
  final Color color;
  const MotionBlurPainter(this.color);

  @override
  void paint(Canvas canvas, Size size) {
    final s = size.shortestSide / 24.0;
    final origin = Offset(
      size.width / 2 - 12.0 * s,
      size.height / 2 - 12.0 * s,
    );
    Offset at(double x, double y) => origin + Offset(x * s, y * s);
    final paint = Paint()
      ..color = color
      ..style = PaintingStyle.stroke
      ..strokeWidth = 2.0 * s
      ..strokeCap = StrokeCap.butt;
    // The ring: a 2-unit stroke on a 4-unit radius, centred at (17, 12).
    canvas.drawCircle(at(17, 12), 4.0 * s, paint);
    // The streaks; two rows broken by a shorter dash further left, which is
    // what makes the mark read as motion rather than a plain arrow.
    const rows = [
      (4.0, 14.0, 8.0),
      (10.0, 13.0, 12.0),
      (8.0, 14.0, 16.0),
      (3.0, 7.0, 12.0),
      (4.0, 5.0, 16.0),
    ];
    for (final (x1, x2, y) in rows) {
      canvas.drawLine(at(x1, y), at(x2, y), paint);
    }
  }

  @override
  bool shouldRepaint(MotionBlurPainter oldDelegate) =>
      oldDelegate.color != color;
}
